presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java

```diff
@@ -0,0 +1,113 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.server;
+
+import com.google.common.hash.Hashing;
+import io.airlift.http.client.HttpRequestFilter;
+import io.airlift.http.client.Request;
+import io.airlift.log.Logger;
+import io.airlift.node.NodeInfo;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jws;
+import io.jsonwebtoken.JwtException;
+import io.jsonwebtoken.JwtParser;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.prestosql.server.security.InternalPrincipal;
+
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+
+import java.security.Principal;
+import java.time.ZonedDateTime;
+import java.util.Date;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static io.airlift.http.client.Request.Builder.fromRequest;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class InternalAuthenticationManager
+        implements HttpRequestFilter
+{
+    private static final Logger log = Logger.get(InternalAuthenticationManager.class);
+
+    private static final String PRESTO_INTERNAL_BEARER = "X-Presto-Internal-Bearer";
+
+    private final Optional<JwtParser> jwtParser;
+    private final Optional<Supplier<String>> jwtGenerator;
+
+    @Inject
+    public InternalAuthenticationManager(InternalCommunicationConfig internalCommunicationConfig, NodeInfo nodeInfo)
+    {
+        Optional<String> sharedSecret = requireNonNull(internalCommunicationConfig, "internalCommunicationConfig is null").getSharedSecret();
+        if (sharedSecret.isPresent()) {
+            byte[] hmac = Hashing.sha256().hashString(sharedSecret.get(), UTF_8).asBytes();
+            this.jwtParser = Optional.of(Jwts.parser().setSigningKey(hmac));
+            String nodeId = nodeInfo.getNodeId();
+            this.jwtGenerator = Optional.of(() -> generateJwt(hmac, nodeId));
+        }
+        else {
+            this.jwtParser = Optional.empty();
+            this.jwtGenerator = Optional.empty();
+        }
+    }
+
+    private static String generateJwt(byte[] hmac, String nodeId)
+    {
+        return Jwts.builder()
+                .signWith(SignatureAlgorithm.HS256, hmac)
+                .setSubject(nodeId)
+                .setExpiration(Date.from(ZonedDateTime.now().plusMinutes(5).toInstant()))
+                .compact();
+    }
+
+    public boolean isInternalRequest(HttpServletRequest request)
+    {
+        return request.getHeader(PRESTO_INTERNAL_BEARER) != null;
+    }
+
+    public Principal authenticateInternalRequest(HttpServletRequest request)
+    {
+        if (!jwtParser.isPresent()) {
+            log.error("Internal authentication in not configured");
+            return null;
```

Comment:
why not throw in this case?

Reply:
or at least `Optional`

Reply:
This is modeled after the other authenticators APIs.